File: src/main/java/aoc/aoc2019/Dec11.java
package aoc.aoc2019;

import static aoc.utils.Utils.check;
import static aoc.utils.Utils.getLines;
import static aoc.utils.Utils.readAnswerAsInt;

import java.util.HashMap;

public class Dec11 {

  public static void main(String[] args) throws Exception {
    task1();
    task2();
  }

  public static void task1() throws Exception {
    HashMap<String, Long> panels = drawPanels(0L);
    System.out.println("Result: " + panels.size());
    check(panels.size(), readAnswerAsInt(1));
  }

  public static void task2() throws Exception {
    HashMap<String, Long> panels = drawPanels(1L);
    System.out.println("Result: " + panels.size());
    printPanels(panels);
  }

  private static HashMap<String, Long> drawPanels(long initial) {

    HashMap<String, Long> panels = new HashMap<>();
    panels.put("0,0", initial);
    int x = 0, y = 0;
    char dir = '^';
    long color, turn;

    IntCode.Program p = IntCode.Program.create(getLines().get(0));
    while (p.status != IntCode.Program.Status.HALT_OK) {
      color = color(panels, pos(x, y));
      p.input.add(color);

      IntCode.execute(p);
      color = p.output.poll();
      turn = p.output.poll();
      panels.put(pos(x, y), color);
      dir = getDir(dir, turn);
      x += dx(dir);
      y += dy(dir);
      // System.out.println("Step: " + dir + " " + x + "," + y + " " + newColor);
    }
    return panels;
  }

  private static long color(HashMap<String, Long> panels, String pos) {
    return panels.containsKey(pos) ? panels.get(pos) : 0L;
  }

  private static int dx(char dir) {
    if (dir == '<') {
      return -1;
    }
    if (dir == '>') {
      return 1;
    }
    return 0;
  }

  private static int dy(char dir) {
    if (dir == '^') {
      return 1;
    }
    if (dir == 'v') {
      return - 1;
    }
    return 0;
  }

  private static String pos(int x, int y) {
    return x + "," + y;
  }

  public static void printPanels(HashMap<String, Long> panels) {
    int minx = 0, maxx = 0, miny = 0, maxy = 0;
    for (String panel : panels.keySet()) {
      minx = Math.min(minx, Integer.parseInt(panel.split(",")[0]));
      maxx = Math.max(maxx, Integer.parseInt(panel.split(",")[0]));
      miny = Math.min(miny, Integer.parseInt(panel.split(",")[1]));
      maxy = Math.max(maxy, Integer.parseInt(panel.split(",")[1]));
    }
    for (int y = maxy; y >= miny; y--) {
      for (int x = maxx; x >= minx; x--) {
        System.out.print(color(panels, pos(x, y)) == 1 ? "*" : " ");
      }
      System.out.println();
    }
  }

  private static char getDir(char dir, long turn) {
    switch (dir) {
      case '^':
        return turn == 1L ? '<' : '>';
      case '>':
        return turn == 1L ? '^' : 'v';
      case '<':
        return turn == 1L ? 'v' : '^';
      case 'v':
        return turn == 1L ? '>' : '<';
      default:
        throw new RuntimeException("CMH");
    }
  }
}

File: src/main/kotlin/aoc/aoc2017/Dec11.kt
package aoc.aoc2017

import aoc.ktutils.check
import aoc.ktutils.readText
import aoc.ktutils.readAnswerAsInt

fun main() {
    check(execute("ne,ne,ne").first, 3)
    check(execute("ne,ne,sw,sw").first, 0)
    check(execute("ne,ne,s,s").first, 2)
    check(execute("se,sw,se,sw,sw").first, 3)
    check(execute("ne,nw").first, 1)
    check(execute("n,n,n,sw").first, 3)
    check(execute("s,s,s,sw").first, 4)
    check(execute("s,sw,sw,sw").first, 4)
    check(execute("s,s,s,nw").first, 3)
    check(execute("s,sw,sw,sw").first, 4)
    check(execute("s,s,s,nw").first, 3)
    execute(readText()).let { println(it.first) ; check(it.first, readAnswerAsInt(1)) }

    check(execute("ne,ne,sw,sw").second, 2)
    execute(readText()).let { println(it.second) ; check(it.second, readAnswerAsInt(2)) }
}

private fun execute(input: String): Pair<Int, Int> {
    val steps = input.split(",")
    var n = 0
    var e = 0
    var dist = 0
    var max = 0
    for (step in steps) {
        when (step) {
            "n" -> {n++}
            "s" -> {n--}
            "ne" -> {n++ ; e++}
            "nw" -> {e--}
            "se" -> {e++}
            "sw" -> {n--; e--}
            else -> throw RuntimeException("CMH: '$step'")
        }
        dist = hexDistance(e, n)
        max = kotlin.math.max(dist, max)
    }
    return Pair(dist, max)
}

fun hexDistance(e: Int, n: Int): Int {
    if (n >= 0 && e >= 0) return Math.max(n, e)
    if (n >= 0 && e < 0) return n + Math.abs(e)
    if (n < 0 && e >= 0) return Math.abs(n) + e
    if (n < 0 && e < 0) return Math.max(Math.abs(n), Math.abs(e))
    throw RuntimeException("CMH: ($e, $n)")
}
